File: basic_dsa_sixty_eight.java
/**
 * Basic DSA Problems - One More Fundamental Question
 * Simple and essential DSA problems for beginners
 */
import java.util.*;

public class basic_dsa_sixty_eight {

    // Problem: Remove Duplicates from Sorted Array (in-place)
    // Return the number of unique elements; first k slots should hold unique elements in order.
    // Example: [1,1,2,2,3] -> k=3, array becomes [1,2,3,...]

    /**
     * Remove duplicates in-place, return count of unique elements
     * Time Complexity: O(n)
     * Space Complexity: O(1)
     */
    public static int removeDuplicates(int[] nums) {
        if (nums == null || nums.length == 0) return 0;
        int k = 1;
        for (int i = 1; i < nums.length; i++) {
            if (nums[i] != nums[k - 1]) {
                nums[k++] = nums[i];
            }
        }
        return k;
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Remove Duplicates from Sorted Array");
        System.out.println("=".repeat(60));

        int[] a1 = { 1, 1, 2, 2, 3 };
        int k1 = removeDuplicates(a1);
        System.out.println("Input: [1,1,2,2,3] -> k = " + k1 + ", unique: " + Arrays.toString(Arrays.copyOf(a1, k1)));

        int[] a2 = { 1, 2, 3 };
        int k2 = removeDuplicates(a2);
        System.out.println("Input: [1,2,3] -> k = " + k2 + ", unique: " + Arrays.toString(Arrays.copyOf(a2, k2)));

        int[] a3 = { 1, 1, 1 };
        int k3 = removeDuplicates(a3);
        System.out.println("Input: [1,1,1] -> k = " + k3 + ", unique: " + Arrays.toString(Arrays.copyOf(a3, k3)));
    }
}

File: basic_dsa_sixty_eight.py
"""
Basic DSA - Subarray Sum Equals K
Return the number of contiguous subarrays whose elements sum to k.
"""


def subarray_sum(nums, k):
    """
    Prefix sum with hash map of prefix frequency.
    If prefix[j] - prefix[i] = k, then sum(nums[i:j]) = k.
    Time Complexity: O(n)
    Space Complexity: O(n)
    """
    count = 0
    prefix = 0
    seen = {0: 1}

    for x in nums:
        prefix += x
        need = prefix - k
        count += seen.get(need, 0)
        seen[prefix] = seen.get(prefix, 0) + 1

    return count


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Subarray Sum Equals K")
    print("=" * 60)

    tests = [
        ([1, 1, 1], 2, 2),
        ([1, 2, 3], 3, 2),
        ([1], 0, 0),
        ([1, -1, 0], 0, 3),
    ]

    for nums, k, expected in tests:
        got = subarray_sum(nums, k)
        print(f"nums={nums}, k={k} -> {got} (expected {expected})")
